### ISSUE TITLE
Self-closing tags are not correctly recognized

### ISSUE THREAD

**Issue by axbender** [htacg/tidy-html5#813]
Version: 5.7.16 (and before)
Given the input
---------------
<some_tag/>
---------------
tidy outputs
---------------------------
<?xml version="1.0"?>
<some_tag /></some_tag>
---------------------------
i.e. it adds a closing tag to an already closed one...

This happens with the attached config file (and only, if output-html is given).
--------------------------------------------------
accessibility-check:         0
add-meta-charset:            no
add-xml-decl:                no
add-xml-space:               no
alt-text:
anchor-as-name:              yes
ascii-chars:                 no
assume-xml-procins:          no
bare:                        no
break-before-br:             no
char-encoding:               utf8
clean:                       yes
coerce-endtags:              yes
css-prefix:                  c
custom-tags:                 no
decorate-inferred-ul:        no
doctype:                     html5
drop-empty-elements:         yes
drop-empty-paras:            yes
drop-proprietary-attributes: yes
enclose-block-text:          no
enclose-text:                no
error-file:
escape-cdata:                no
escape-scripts:              no
fix-backslash:               yes
fix-bad-comments:            auto
fix-style-tags:              yes
fix-uri:                     yes
force-output:                no
gdoc:                        no
gnu-emacs:                   yes
hide-comments:               no
indent-attributes:           no
indent-cdata:                no
indent-spaces:               3
indent-with-tabs:            no
indent:                      yes
input-encoding:              utf8
input-xml:                   yes
join-classes:                no
join-styles:                 yes
keep-tabs:                   no
keep-time:                   no
literal-attributes:          no
logical-emphasis:            no
lower-literals:              yes
markup:                      yes
merge-divs:                  auto
merge-emphasis:              yes
merge-spans:                 auto
mute-id:                     no
mute:                        STRING_MUTING_TYPE
ncr:                         yes
new-blocklevel-tags:
new-empty-tags:
new-inline-tags:
new-pre-tags:
newline:                     CRLF
numeric-entities:            no
omit-optional-tags:          no
output-bom:                  no
output-encoding:             utf8
output-file:
output-html:                 yes
output-xhtml:                no
output-xml:                  no
preserve-entities:           no
priority-attributes:
punctuation-wrap:            no
quiet:                       yes
quote-ampersand:             yes
quote-marks:                 no
quote-nbsp:                  yes
repeated-attributes:         keep-last
replace-color:               no
show-body-only:              no
show-errors:                 999
show-filename:               no
show-info:                   no
show-meta-change:            no
show-warnings:               yes
skip-nested:                 yes
sort-attributes:             none
strict-tags-attributes:      no
tab-size:                    8
tidy-mark:                   yes
uppercase-attributes:        no
uppercase-tags:              no
vertical-space:              no
warn-proprietary-attributes: yes
word-2000:                   no
wrap-asp:                    yes
wrap-attributes:             no
wrap-jste:                   yes
wrap-php:                    yes
wrap-script-literals:        no
wrap-sections:               yes
wrap:                        68
write-back:                  no

**Comment by geoffmcl** [comment]
@axbender thanks for the issue... this seems a **question** between `XML` **and** `HTML`... a conversion...

I can **repeat** the output with `-xml`, that is your `input-xml: yes`, **and** `-ashtml`, your `output-html: yes`, options, on your input... so do not need **all** the config... but what does this indicate?

You are exactly correct! `tidy` has no understanding of `self-closing` tags... maybe the [HTML 5.2 REC](https://www.w3.org/TR/html5/syntax.html) might help in that...

I think `tidy` can **not** really be used to convert `xml` to `html`... that is **not** in its purpose, aims... although it does its best... but probably **fails** in many instances... is that what we are seeing here?

If **yes**, then I am not sure `tidy` **should**, or **could**, be improved in this regard. What do others think?

Also need to search for other related issues... any pointers appreciated...

One tool I know, to convert `XML` to `HTML`, is `xsltproc`... you supply a `style.xsl` for your `xml`, include `<xsl:output method="html"/>` in this, and process the `xml` you have... get html... there may be others...

We use this to generate the `tidy.1` man file, and the [QuickRef](http://api.html-tidy.org/tidy/quickref_next.html) `HTML`, from a tidy `XML` output... see `man/tidy1.xsl.in` in the source for part of this... and the `CMakeLists.txt` for the generation steps...

Again, I do not think `tidy` is a **good** tool for such **conversions**... but maybe it should...

Look forward to further feedback, comments, patches, PR, ... to test... thanks...

**Comment by balthisar** [comment]
Lacking further feedback from the bug filer, I will close this.

**Comment by jscott0** [comment]
> @balthisar Lacking further feedback from the bug filer, I will close this.

I'm not the bug filer, but I'd like to report my user experience where this issue comes into play. I am creating documents in XHTML+RDFa and CSS Paged Media and plan to possibly use even more rich XML features in the future such as XInclude. When that time arises, I'll probably use `libxml2` or something similar to do the post-processing, but for right now I have a valid XHTML document.

Unfortunately I've not been able to find a tool that converts XHTML to HTML except Tidy which gets _very_ close, except for trying to close `<br/>` and thus making two line breaks appear instead of one. An alternative I tried was LibreOffice via <code>soffice  --convert-to html ./foo.xhtml</code>.
This works by importing the XHTML as a LibreOffice Writer document with the XHTML importer and then exporting with the HTML exporter. This serializes and deserializes the document into LibreOffice's internal representation, and in practice it seems to mangle or discard styles. I'm not sure LibreOffice is able to handle much CSS.

> @geoffmcl I think tidy can not really be used to convert xml to html... that is not in its purpose, aims... although it does its best... but probably fails in many instances... is that what we are seeing here?

It does a lot better than almost any other tool, and it's in a better _position_ to, I think. In every other respect Tidy is made to handle XML, HTML, and be very permissive about the quality of its input. One quirk that tools blissfully unaware of real-world HTML issues gloss over is with my document's CSS containing the line: <pre><code>@namespace url("http://www.w3.org/1999/xhtml");</code></pre>

Conversion of XHTML to HTML may or may not preserve the `xmlns="http://www.w3.org/1999/xhtml"` declaration on my top-level `<html>` element, but regardless, user agents that don't intend to support XML or XHTML at all <a href="https://github.com/Kozea/WeasyPrint/">such as WeasyPrint</a> will expect the <code>@namespace</code> rule to be gone. Even if the <code>xmlns=...</code> bit is preserved, WeasyPrint and like pure-HTML user agents will be confused by this directive. This is a delicate detail that an XSLT processor can't handle on its own, but it's just barely in the realm of Tidy. What other project gets as deep in the details of real-world HTML documents, the quirks they may have, and smooths them out?

> tidy has no understanding of self-closing tags... maybe the [HTML 5.2 REC](https://www.w3.org/TR/html5/syntax.html) might help in that...

The syntax for self-closing tags is a standard XML feature, and—I'd have to double-check—I think the HTML spec requires the self-closing notation to be permitted and handled correctly. A good enough hack for the interim might be—if a self-closing tag like `<br/>` is seen—to give it the benefit of the doubt, and pass it through to the output exactly as-is.

In conclusion I would like to see this issue be re-opened: I don't think anyone contests that this is a bug. Pragmatically, Tidy appears to be the best tool for the task I have at hand: XHTML and HTML differences are syntactic quirks much like the kind Tidy deals with in cleaning up HTML documents normally. Suggesting that Tidy understand the CSS `@namespace` rule is a bit of a reach, I'll admit (unless I can submit a patch someday), and I don't mean to tell you what to do as someone who's not yet a contributor, but it seems like destiny that this belongs on Tidy's roadmap and handling XHTML/HTML quirks is a function it could assimilate into.
